lib/http-bridge/src/Cardano/Wallet/HttpBridge.hs

```diff
@@ -0,0 +1,156 @@
+import Cardano.Launcher
+    ( ProcessHasExited (..), installSignalHandlers )
+import Cardano.Wallet
+    ( WalletLayer )
+import Cardano.Wallet.Api.Server
+    ( Listen (..) )
+import Cardano.Wallet.DaedalusIPC
+    ( daedalusIPC )
+import Cardano.Wallet.DB
+    ( DBLayer )
+import Cardano.Wallet.HttpBridge.Compatibility
+    ( HttpBridge )
+import Cardano.Wallet.HttpBridge.Environment
+    ( KnownNetwork (..) )
+import Cardano.Wallet.HttpBridge.Network
+    ( ErrStartup (..), HttpBridgeBackend (..) )
+import Cardano.Wallet.HttpBridge.Primitive.Types
+    ( Tx )
+import Cardano.Wallet.Network
+    ( NetworkLayer (..) )
+import Cardano.Wallet.Network.Ports
+    ( PortNumber )
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( KeyToAddress )
+import Cardano.Wallet.Primitive.AddressDerivation.Sequential
+    ( SeqKey )
+import Cardano.Wallet.Primitive.AddressDiscovery.Sequential
+    ( SeqState )
+import Cardano.Wallet.Primitive.Types
+    ( Block )
+import Control.Concurrent
+    ( threadDelay )
+import Control.Concurrent.Async
+    ( race_ )
+import Control.Monad
+    ( forever )
+import Data.Function
+    ( (&) )
+import Data.Maybe
+    ( fromMaybe )
+import Data.Text
+    ( Text )
+import Data.Text.Class
+    ( ToText (..) )
+import Network.Wai.Handler.Warp
+    ( setBeforeMainLoop )
+import System.Exit
+    ( ExitCode (..) )
+
+import qualified Cardano.BM.Configuration.Model as CM
+import qualified Cardano.Wallet as Wallet
+import qualified Cardano.Wallet.Api.Server as Server
+import qualified Cardano.Wallet.DB.Sqlite as Sqlite
+import qualified Cardano.Wallet.HttpBridge.Network as HttpBridge
+import qualified Cardano.Wallet.HttpBridge.Transaction as HttpBridge
+import qualified Data.Text as T
+import qualified Network.Wai.Handler.Warp as Warp
+
+-- | The @cardano-wallet-http-bridge@ main function.
+serveWallet
+    :: forall t k n s a. (t ~ HttpBridge n, s ~ SeqState t, k ~ SeqKey)
+    => (KeyToAddress t k, KnownNetwork n)
+    => (CM.Configuration, Trace IO Text)
+    -- ^ Logging config.
+    -> Maybe FilePath
+    -- ^ Database file.
+    -> Listen
+    -- ^ HTTP API Server port.
+    -> HttpBridgeBackend
+    -- ^ Whether and how to launch or use the node backend.
+    -> Maybe (PortNumber -> NetworkLayer IO Tx (Block Tx) -> Sqlite.SqliteContext -> DBLayer IO s t k -> WalletLayer s t k -> IO a)
+    -- ^ Optional operation to run under the wallet server (e.g. test suite).
+    -> IO ExitCode
+serveWallet (cfg, tr) dbFile listen bridge mAction = do
+    installSignalHandlers tr
+    logInfo tr "Wallet backend server starting..."
+    logInfo tr $ "Node is Http-Bridge on " <> toText (networkVal @n)
+    withDBLayer $ \(ctx, db) ->
+        HttpBridge.withNetworkLayer @n tr bridge $ \_bridgePort -> \case
+            Right nl -> do
+                wl <- newWalletLayer db nl
+                withServer wl $ \apiPort -> do
+                    let action = fromMaybe (\_ _ _ _ _ -> forever $ threadDelay maxBound) mAction
```

Comment:
I am sure we can do better here :grimacing: 
What about passing the `maybe action` down to `withServer` and have a pattern match here:

- If there's an action, we go for the double-race
- If not, then we only spawn the wallet server until it dies.

Reply:
Yeah it seems like whenever i start using `Maybe (IO a)` the code gets ugly.
I have fixed it to pass the `maybe action` down.